wgengine/tsdns/tsdns.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package tsdns provides a Resolver struct capable of resolving
+// domains on a Taiscale network.
+package tsdns
+
+import (
+	"encoding/binary"
+	"errors"
+	"strconv"
+
+	"github.com/tailscale/wireguard-go/device"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/packet"
+)
+
+const (
+	// MaxQuerySize is the maximal size of a Magic DNS query.
+	MaxQuerySize = 512
+	// MaxResponseSize is the maximal size of a Magic DNS response.
+	MaxResponseSize = 512
+)
+
+const (
+	ipOffset      = device.MessageTransportHeaderSize
+	dnsDataOffset = ipOffset + packet.UDPDataOffset
+
+	bufferSize = dnsDataOffset + MaxResponseSize
+)
+
+// Resolver is a
+type Resolver struct {
+	logf logger.Logf
+
+	// ip is the IP on which the resolver is listening (default 100.100.100.100).
+	ip packet.IP
+	// port is the port on which the resolver is listening (default 53).
+	port uint16
+
+	// responseBuffer to avoid graticious allocations.
+	responseBuffer [bufferSize]byte
+}
+
+// NewResolver constructs a resolver with default parameters.
+func NewResolver(logf logger.Logf) *Resolver {
+	return &Resolver{
+		logf: logf,
+		ip:   packet.IP(binary.BigEndian.Uint32([]byte{100, 100, 100, 100})),
+		port: uint16(53),
+	}
+}
+
+// AcceptsPacket determines if the given packet is a DNS request
+// directed to this resolver (by ip and port).
+// We also require that UDP be used to simplify things for now.
+func (r *Resolver) AcceptsPacket(in *packet.QDecode) bool {
+	return in.DstIP == r.ip && in.DstPort == r.port && in.IPProto == packet.UDP
+}
+
+var (
+	ErrIncomplete       = errors.New("query incomplete")
```

Comment:
These can probably be unexported. I don't think we need them to be part of the package API. The caller won't be taking different actions on these exact errors. If/when that need arises we can export as needed.